### PR TITLE
Enable silent login for api requests

### PR DESCRIPTION
**Pull request by Bibob7**
## Description

I added a new flag called `api-mode`, which enables silent login if set to true, for api requests that expects json content in the response.

## Motivation and Context

[#802 Oauth2 Proxy should also be able to perform silent login when Accept: application/json header is set](https://github.com/oauth2-proxy/oauth2-proxy/issues/802)
## How Has This Been Tested?

I added tests that are covering the new implemented feature.
Additionally, I tested it manually with my own implemented backend api.

By default the `api-mode` flag is set to false, which means that the oauth2-proxy behaves the same as bevor.

## Checklist:

- [X] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by JoelSpeed** [comment]
Is the intention here that we disable the normal authentication flow and have the proxy as just an authentication proxy that checks for a valid token and then allows requests through?

**Comment by Bibob7** [comment]
I would see the silent Login with `prompt=none` still as the normal authentication flow. But apart from that, yes.

**Comment by NickMeves** [comment]
> Is the intention here that we disable the normal authentication flow and have the proxy as just an authentication proxy that checks for a valid token and then allows requests through?

I feel like this is asked for a lot.  I know I've theorized about the potential of a purpose built provider ONLY meant for machine users via bearer headers in another thread related to this.

I wonder if it would just be a provider with only `CreateSessionStateFromBearerToken` with the other provider methods overriding the default and throwing an error if the normal codepath that results in human OAuth2 flows is attempted?

File: oauthproxy.go [hunk 1]
```diff
@@ -393,6 +395,14 @@ func (p *OAuthProxy) RobotsTxt(rw http.ResponseWriter) {
 
 // ErrorPage writes an error response
 func (p *OAuthProxy) ErrorPage(rw http.ResponseWriter, code int, title string, message string) {
+	if p.APIMode {
+		rw.Header().Set("Content-Type", applicationJSON)
```
Comment (NickMeves) [inline oauthproxy.go:399]: Should we just extend the functionality implemented in `ErrorJSON` with the improvements here and call that here to unify things?

File: oauthproxy.go [hunk 2]
```diff
@@ -851,13 +861,13 @@ func (p *OAuthProxy) Proxy(rw http.ResponseWriter, req *http.Request) {
 
 	case ErrNeedsLogin:
 		// we need to send the user to a login screen
-		if isAjax(req) {
+		if isAjax(req) && !p.APIMode {
 			// no point redirecting an AJAX request
 			p.ErrorJSON(rw, http.StatusUnauthorized)
```
Comment (NickMeves) [inline oauthproxy.go:866]: If we are extending the JSON payload to have more context on the error, I think we should add that message here potentially to align everything?

File: docs/configuration/configuration.md [hunk 3]
```diff
@@ -23,6 +23,7 @@ An example [oauth2-proxy.cfg]({{ site.gitweb }}/contrib/oauth2-proxy.cfg.example
 | Option | Type | Description | Default |
 | ------ | ---- | ----------- | ------- |
 | `--acr-values` | string | optional, see [docs](https://openid.net/specs/openid-connect-eap-acr-values-1_0.html#acrValues) | `""` |
+| `--api-mode` | bool | set to true if upstream is a backend api that expects and responds json content
```
Comment (NickMeves) [inline docs/configuration/configuration.md:26]: I think this is better described from the client point of view -- if the client expects JSON payloads in the response, right?

**Pull request closed without merging**